agent/taskresource/credentialspec/credentialspec.go

```diff
@@ -0,0 +1,556 @@
+func (cs *CredentialSpecResource) Create() error {
+	// To fail fast, check execution role first
+	executionCredentials, ok := cs.credentialsManager.GetTaskCredentials(cs.getExecutionCredentialsID())
+	if !ok {
+		// No need to log here. managedTask.applyResourceState already does that
+		err := errors.New("credential spec resource: unable to find execution role credentials")
+		cs.setTerminalReason(err.Error())
+		return err
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	for credSpecStr, _ := range cs.requiredCredentialSpecs {
+		credSpecSplit := strings.SplitAfterN(credSpecStr, "credentialspec:", 2)
+		credSpecValue := credSpecSplit[1]
+
+		if strings.HasPrefix(credSpecValue, "file://") {
+			dockerHostconfigSecOptCredSpec := strings.Replace(credSpecStr, "credentialspec:", "credentialspec=", 1)
+			cs.updateCredSpecMapping(credSpecStr, dockerHostconfigSecOptCredSpec)
+
+			return nil
+		}
+
+		parsedARN, err := arn.Parse(credSpecValue)
+		if err != nil {
+			return err
+		}
+
+		parsedARNService := parsedARN.Service
+		if parsedARNService == "s3" {
+			s3ResourceARN := parsedARN
+			s3CredSpecValue := credSpecValue
+
+			bucket, key, err := s3.ParseS3ARN(s3CredSpecValue)
+			if err != nil {
+				return err
+			}
+
+			s3Client, err := cs.s3ClientCreator.NewS3ClientForBucket(bucket, cs.region, iamCredentials)
+			if err != nil {
+				return errors.Wrapf(err, "unable to initialize s3 client for bucket %s", bucket)
+			}
+
+			resourceBase := filepath.Base(s3ResourceARN.Resource)
+			localCredSpecFilePath := fmt.Sprintf("%s/s3_%s_%s.json", resourceDir, cs.taskARN, resourceBase)
+
+			err = cs.writeS3File(func(file oswrapper.File) error {
+				return s3.DownloadFile(bucket, key, s3DownloadTimeout, file, s3Client)
+			}, localCredSpecFilePath)
+			if err != nil {
+				return errors.Wrapf(err, "unable to download s3 file %s from bucket %s", key, bucket)
+			}
+
+			dockerHostconfigSecOptCredSpec := fmt.Sprintf("credentialspec=file://%s", localCredSpecFilePath)
+			cs.updateCredSpecMapping(credSpecValue, dockerHostconfigSecOptCredSpec)
+
+		} else if parsedARNService == "ssm" {
+			ssmResourceARN := parsedARN
+
+			ssmClient := cs.ssmClientCreator.NewSSMClient(cs.region, iamCredentials)
+
+			ssmParam := filepath.Base(ssmResourceARN.Resource)
+			ssmParams := []string{ssmParam}
+
+			ssmParamMap, err := ssm.GetParametersFromSSM(ssmParams, ssmClient)
+			if err != nil {
+				return err
+			}
+
+			ssmParamData := ssmParamMap[ssmParam]
+
+			localCredSpecFilePath := fmt.Sprintf("%s/ssm_%s_%s.json", resourceDir, cs.taskARN, ssmParam)
+
+			err = cs.writeSSMFile(ssmParamData, localCredSpecFilePath)
+			if err != nil {
+				return err
+			}
+
+			dockerHostconfigSecOptCredSpec := fmt.Sprintf("credentialspec=file://%s", localCredSpecFilePath)
+			cs.updateCredSpecMapping(credSpecValue, dockerHostconfigSecOptCredSpec)
+
+		} else {
+			return errors.New("unsupported credential spec ARN dependency")
```

Comment:
If we call out currently only support S3 and SSM in the error message, it will be easier for customers to know what is wrong.